### PR TITLE
Controlled MCDA

### PR DESCRIPTION
**Pull request by Flaminietta**
The major changes introduced by this PR are:

1. the **_user can now decide_** on a full sensitivity analysis (all pairs norm/agg functions are used and compared) or only a fixed pair norm/agg functions is used in the calculation of the scores.
2. the **_configuration.json_** file has been reorganized for an easier parameter settings.
3. a more **_context-sensitive renaming_** happened (see variability/robustness analysis definitions in the README file).

Unit tests pass all. I performed successfully the 8 different possible end-to-end tests, namely:

1. no  sensitivity  - no robustness
2. yes sensitivity - no robustness
3. no  sensitivity  - yes robustness - on indicators
4. no sensitivity   -  yes robustness - on all weights
5. no sensitivity   -  yes robustness - on one weight at the time
6. yes sensitivity -  yes robustness - on indicators
7. yes sensitivity  -  yes robustness - on all weights
8. yes sensitivity  -  yes robustness - on one weight at the time

(see attached figures, all made by using the 12 alternatives of the "climchalp" dataset).

@mspada & @kapil-agnihotri please make a review of the code.

@mspada please use another test case for a double round of end-to-end tests (also the ones that should exit with an error message). Additionally (and optionally), you might want to think of a more elegant solution for the hard-coded column names in the aggregation functions (now they are explicitly assigned if `method==None` because the logic used for `col_names_method` cannot work correctly in this case due to the funneling logic in the if-blocks).

@thorsten-reitz & @Somakanthan you are just observers. Any comment is welcome.

### CODE REVIEW

**Comment by Flaminietta** [comment]
These are some of the plots of my end-to-end tests.
MCDA_norm_scores_no_var
MCDA_norm_scores_no_var
MCDA_rand_indicators_rough_scores
MCDA_rand_weights_rough_scores

File: tests/unit_tests/test_mcda_without_robustness.py
Comment (kapil-agnihotri) [inline tests/unit_tests/test_mcda_without_robustness.py:15]: Is this ` "input_matrix_path": "/path/to/input_matrix.csv",` something that the user needs to update before running these tests?
If so then probably here as well will it be possible to use relative path so that user won't have to update it at several places(in below three methods)?

File: tests/unit_tests/test_mcda_with_robustness.py
Comment (kapil-agnihotri) [inline tests/unit_tests/test_mcda_with_robustness.py:79]: It is slightly confusing to me if `_w_` is `with` or `without`.
If I see it at the place of initialisation, then it is clear that it is `with` but reading it down in the code leads to confusion.

File: tests/unit_tests/test_mcda_with_robustness.py
Comment (kapil-agnihotri) [inline tests/unit_tests/test_mcda_with_robustness.py:9]: Maybe I missed before, is there any convention to use just `camelCase` for class names and `_` in the name of test classes?
Can we have same formatting for class names irrespective of test or a normal class? Someone might confuse test class names with `_` to a variable name.

File: configuration.json
Comment (Flaminietta) [reply configuration.json:2]: @mspada After a discussion with Kapil, we agreed to leave the relative paths as they are for now. The reason is that either the user gives the relative paths in the configuration file (like now), or she/he has to copy the input data somewhere specified in the readme file (like it would be if we use absolute paths).

**Review by mspada — CHANGES_REQUESTED** [review]
I reviewed the codes and I agree with the comments from Kapil. 
On top of that, I also run the End-To-End Test for all possible combinations between sensitivity and robustness, as you did, for a case study of a former student of mine. Everything run as it should. However, I had to modify some of the indicators, since it was not running for some of the cases. In this context, I realised that:

- If an indicator has negative numbers (and this can happen with real case datasets) the only WS and minimum are working the other 2 no. There are 2 ways, either we write in the README (and on the paper) that the Indicators should have positive numbers or we put a check in the normalisation script, which look at  negative values and if this is so, shift the values in order to get the lowest to 0

- Also if the std dev is larger than the mean, this could produce negative numbers, which should then be shifted as well, otherwise it will produces zeros... not trivial

- If an Indicator has a 0 value, this is not accepted, why? I have quite some examples that shows we could have some indicators, which have 0 for some of the alternatives

On top of these, I have also some additional comments, which should be addressed (principally the first one):

- An std deviation column when the indicator is "exact" should not be requested

- csv files are not generated, but rather .pkl ones when "on_single_weights is selected", is there a reason for that? To many data for a csv file?

- Figure names are the same in some cases, as for example for "no sensitivity - no robustness" and "yes sensitivity - no robustness". This is a problem when the user would like to run the ProMCDA for different cases (keeping the result directory unchanged)

File: mcda/configuration/config.py
Comment (Flaminietta) [reply mcda/configuration/config.py:96]: Lines 95-101 correctly define the two properties `sensitivity` and `robustness` as both are present in the configuration file.

File: tests/unit_tests/test_mcda_without_robustness.py
Comment (Flaminietta) [reply tests/unit_tests/test_mcda_without_robustness.py:16]: This is only a placeholder and tests run without updating the path. The point is that the` input_matrix_path` is never used within the tests and all needed test matrices are given through functions like `get_input_matrix`.

File: tests/unit_tests/test_mcda_with_robustness.py
Comment (Flaminietta) [reply tests/unit_tests/test_mcda_with_robustness.py:79]: Solved. I have refactored all in `mcda_with_robustness`.

File: tests/unit_tests/test_mcda_with_robustness.py
Comment (Flaminietta) [reply tests/unit_tests/test_mcda_with_robustness.py:9]: Apparently, in Python one can use underscores to separate words in class names and in test class names, following the convention known as "snake_case", as well as the CamelCase convention. Functions instead should be defined by using the "snake_case" convention only. Classes and functions cannot easily be confused because classes are always preceded by the word `class`.

**Comment by Flaminietta** [comment]
@kapil-agnihotri I have addressed all your comments to this moment (not Matteo's yet). Additionally, I've added:

-  a new and updated version of the README file (with a logo for the package, badges, and an updated text);
- a check in GitHub actions: pytest build.

Happy if you can have a second look.

**Comment by Flaminietta** [comment]
@mspada,  you wrote:

> If an indicator has negative numbers (and this can happen with real case datasets) the only WS and minimum are working the other 2 no. There are 2 ways, either we write in the README (and on the paper) that the Indicators should have positive numbers or we put a check in the normalisation script, which look at negative values and if this is so, shift the values in order to get the lowest to 0

Both solutions are fine with me. My preference though would be to modify the code to work with negative values too. However, I am doubtful that simply shifting the negative values to 0 is fully correct. I'd rather believe that a rescaling of the entire set of values to a [0,>0] scale might be more appropriate. My rationale is: if one simply shifts the negative values into 0s, then the relative differences of the values are not kept and there is a loss of information. Unless negative values are not meaningful (uninformative) and one needs simply to truncate the set of values. What is your comment on this?

**Comment by Flaminietta** [comment]
@mspada 

> Also if the std dev is larger than the mean, this could produce negative numbers, which should then be shifted as well, otherwise it will produce zeros... not trivial

If there is a case with std values larger than the means, I am even not sure if it makes sense to use them at all for further investigations before having understood and clarified the nature and causes of such variability in the data.

If we really want to keep this case (and not exit the calculation with a warning), then we should consider special data transformations. We might consider data transformation techniques to reduce the variability, such as log transformation or standardization. What do you suggest here?

**Comment by Flaminietta** [comment]
@mspada 

> If an Indicator has a 0 value, this is not accepted, why? I have quite some examples that shows we could have some indicators, which have 0 for some of the alternatives

This check is happening in the target normalization function. I don't remember why we agreed to code the function in this way, but the reason seems to avoid a division per 0 (if 0 happens to be the minimum value in the set). What would be the new strategy if 0 is the minimum value? In that case, I add an epsilon?

    ```
def target(self, feature_range: tuple) -> pd.DataFrame():
        """
        Normalizes indicators using the scaling method target.
        :return: pd.DataFrame() of same shape as the input
        """

        original_shape = self._input_matrix.shape

        pol = Normalization._cast_polarities(self)
        ind_plus = pol[0]
        ind_minus = pol[1]
        indicators_plus = pol[2]
        indicators_minus = pol[3]

        if self._input_matrix.values.all() > 0:
            if (feature_range == (0, 1)):
                indicators_scaled_target_plus = indicators_plus / indicators_plus.max(axis=0)  # for + polarity
                indicators_scaled_target_minus = 1 - indicators_minus / indicators_minus.max(axis=0)  # for - polarity
            else:
                indicators_scaled_target_plus = indicators_plus / indicators_plus.max(axis=0) * (
                            1 - 0.1) + 0.1  # for + polarity
                indicators_scaled_target_minus = (1 - indicators_minus / indicators_minus.max(axis=0)) * (
                            1 - 0.1) + 0.1  # for - polarity
        else:
            raise ValueError('Indicators in the input matrix should have all positive values, larger than 0.')
```

**Comment by Flaminietta** [comment]
@mspada 

> csv files are not generated, but rather .pkl ones when "on_single_weights is selected", is there a reason for that? To many data for a csv file?

Yes, in the case one explores the randomness of one weight at the time, the different results are stored in a number of tables equal to the number of indicators (each one representing the results for a particular stocastically sampled weight). For that reason, I cannot use a simple CSV table and decided to use a pickle file (that contains all my tables and can be easily opened in Python). They serialize my Python dictionary object.

**Comment by Flaminietta** [comment]
@mspada @kapil-agnihotri 

> Figure names are the same in some cases, as for example for "no sensitivity - no robustness" and "yes sensitivity - no robustness". This is a problem when the user would like to run the ProMCDA for different cases (keeping the result directory unchanged)

The idea I had is that the user changes the output directory for different tests/simulations. If not, not only the figure names are a problem, but all the output names (of the tables, of the configuration file, etc.).

Let's rethink the strategy here: maybe an option can be to simplify all names to the minimum but add a time stamp to them by using `datetime. now()`. For example:

configuration_datetime.json
rough_scores_datetime.png
rough_scores_datetime.csv
etc.

In this example, the time stamp unifies all outputs belonging to a specific simulation and the configuration helps to recall what settings have been used. We leave to the user her preference to have a single or multiple output directories. What do you think?

**Comment by mspada** [comment]
> @mspada,  you wrote:
> 
> 
> 
> > If an indicator has negative numbers (and this can happen with real case datasets) the only WS and minimum are working the other 2 no. There are 2 ways, either we write in the README (and on the paper) that the Indicators should have positive numbers or we put a check in the normalisation script, which look at negative values and if this is so, shift the values in order to get the lowest to 0
> 
> 
> 
> Both solutions are fine with me. My preference though would be to modify the code to work with negative values too. However, I am doubtful that simply shifting the negative values to 0 is fully correct. I'd rather believe that a rescaling of the entire set of values to a [0,>0] scale might be more appropriate. My rationale is: if one simply shifts the negative values into 0s, then the relative differences of the values are not kept and there is a loss of information. Unless negative values are not meaningful (uninformative) and one needs simply to truncate the set of values. What is your comment on this? Can you always say what case is realistic to come across in MCDA applications with: unrealistic or realistic (but untreatable) negative values?

@Flaminietta, I meant rescaling not just shifting the negative values ;-). I guess, it would be better to have it implemented directly in ProMCDA, rather than leave the manual work to the user.

**Comment by mspada** [comment]
> @mspada 
> 
> 
> 
> > If an Indicator has a 0 value, this is not accepted, why? I have quite some examples that shows we could have some indicators, which have 0 for some of the alternatives
> 
> 
> 
> This check is happening in the target normalization function:
> 
> 
> 
> ```
> 
>         if self._input_matrix.values.all() > 0:
> 
>             if (feature_range == (0, 1)):
> 
>                 indicators_scaled_target_plus = indicators_plus / indicators_plus.max(axis=0)  # for + polarity
> 
>                 indicators_scaled_target_minus = 1 - indicators_minus / indicators_minus.max(axis=0)  # for - polarity
> 
>             else:
> 
>                 indicators_scaled_target_plus = indicators_plus / indicators_plus.max(axis=0) * (
> 
>                             1 - 0.1) + 0.1  # for + polarity
> 
>                 indicators_scaled_target_minus = (1 - indicators_minus / indicators_minus.max(axis=0)) * (
> 
>                             1 - 0.1) + 0.1  # for - polarity
> 
>         else:
> 
>             raise ValueError(Indicators in the input matrix should have all positive values, larger than 0.)
> 
> ```
> 
> 
> 
> I don't remember why we agreed to code the function in this way, but the reason seems to avoid a division per 0 (if 0 happens to be the minimum value in the set). What would be the new strategy if 0 is the minimum value? In that case, I add an epsilon?
> 
> 

@Flaminietta I do not recall why this has been implemented. In any case, I'm also not sure where you will go to divide by 0 in one of the implemented normalization methods. Therefore, I do not see what problem we could have by input raw data with 0 values. We should remove this check

**Comment by mspada** [comment]
> @mspada 
> 
> 
> 
> > Also if the std dev is larger than the mean, this could produce negative numbers, which should then be shifted as well, otherwise it will produce zeros... not trivial
> 
> 
> 
> If there is a case with std values larger than the means, I am even not sure if it makes sense to use them at all for further investigations before having understood and clarified the nature and causes of such variability in the data.
> 
> 
> 
> If we really want to keep this case (and not exit the calculation with a warning), then we should consider special data transformations. We might consider data transformation techniques to reduce the variability, such as log transformation or standardization. What do you suggest here?

@Flaminietta in this case I completely agree with you. In my opinion we should add a warning and exit when the std is larger than the means.

**Review by mspada — APPROVED** [review]
Thanks @Flaminietta, let me think about the rescaling once more (as you stated on slack) ;-)

**Comment by Flaminietta** [comment]
@mspada and  @kapil-agnihotri, I've addressed all your comments now. I think that we need a second-round review from your side before merging. This version of the code is the one we want to finally release on PyPI. 

Short recap on the most important points:

- negative indicator values are now allowed as input: they are rescaled through a minmax rescaler into [0,1].
   @mspada, my idea is that a linear normalization, which keeps the order and relative proximity of the data points, 
   does not cause any artifact when later the values go through another normalization. Do we need to test it with some
   examples to be on the safe side?
- indicators with values = 0 are now also accepted as input.
- if any standard deviation for any indicator has a value larger than the relative mean (case of high variability of data), 
  a warning is introduced. The user can decide to stop or continue the calculations. In this case, the negative values 
  created by sampling will be rescaled as for point 1.
- the readme is updated, @mspada please review as part of the paper too
- all output file are associated with a time stamp
- the input matrix with uncertainty does not need std columns for indicators associated with an "exact" pdf.
   Please, carefully review and test this new functionality in particular. I have seen some differences in results obtained with 
   the previous version of the code. At this point, I cannot see any mistake in the logic implemented now.

**Comment by Flaminietta** [comment]
2023-11-08_10-42-51_MCDA_rough_scores
MCDA_rand_indicators_rough_scores

The first and second plots come from old and new version of the code, same input. There are some differences.

**Comment by mspada** [comment]
@Flaminietta I just made some bugfixes:

- "mcda_with_robustness.py" corrected for how the parameters were given to the uniform distribution
- "mcda_with_robustness.py" and "utils.py" corrected for the Poisson PDF, since it should be treated as an exact
- "utils.py" corrected for the uniform distribution, since it should not be checked if the std is larger than the mean, since the mean is in reality the lowest value and the std the highest defining the distribution.
- "mcda_run.py" corrected for the check of the numbers of Monte-Carlo runs, since when < 1000 the check was looking at a wrong feature name

**Comment by Flaminietta** [comment]
Here we go again @mspada, I hope this is the last step before merging!
This last commit:

- corrects the function that checks parameter 1 and parameter 2 in the matrix with uncertainties;
- corrects also mcda_run_with_robustness;
- refactors a couple of names;
- updates the README (e.g., how to build an input matrix with uncertainty among the rest);
- add some extra unit tests;
- aligns the mcda_run.

Please have a final look into the code and in particular challenge it by running the code with some tests where you play with wrong parameters 1 and 2 in the input matrix.

**Review by mspada — APPROVED** [review]
From my side we are ready to merge. The only thing missing is the updated Logo, which should come in the next days

**Pull request merged**
